### PR TITLE
[Name lookup] Introduce a request for "extended nominal type decl"

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce `ExtensionDecl::getExtendedNominal()` to provide the nominal
type declaration that the extension declaration extends. Move most
of the existing callers of the callers of `getExtendedType()` over to
`getExtendedNominal()`, because they don’t need the full type information.

Introduce a request to back `ExtensionDecl::getExtendedNominal()`, which 
uses the type-decl-resolution path that doesn't go through a `Type`. Use that
to reduce the dependence of the early "extension binding" pass (which associates
extension declarations with nominal types) on the type checker.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by slavapestov** [comment]
The line notes are messed up again, gah!

Basically I’m confused why we ever call bindExtension() directly - are there invalid nested extensions that do lookup into other extensions during binding or something?

**Comment by DougGregor** [comment]
(Writing here because my rebase made line-specific commentary hard to follow)

@slavapestov I confused the issue because I effectively added another level of "binding" for extensions without giving it a name. Here are the three effective "levels":

* Determining the nominal type declaration to which the extension is bound, which makes the extension visible to `getExtensions()`. This is what we do in the worklist-based early type checking pass.
* Determining the type to which the extension is bound, as an actual `Type`, which lets us talk about the `Self` type in context. The calls through `LazyResolver::bindExtension()` do this.
* Computing a full generic signature for the extension. The calls through `LazyResolver::resolveExtension()`. do this.

Prior to this pull request, the first two levels were the same. This pull request splits them apart, creates a request for the former, and scatters `bindExtension()` calls around where clients want the type information from the second level. Request-ifying the second level will eliminate the `bindExtension()` calls, but for now I'd like them to be there because it's a good forcing function for moving these clients over to `getExtendedNominal()`, which is cheaper and doesn't require the type checker.

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/18490 removed the `getAllConformances()` call that broke the Linux build, so now I'll try that again.

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS
